yt/yt/ytlib/api/native/client_queries.cpp

```diff
@@ -38,6 +40,300 @@ using namespace NYTree;
+    if (!error.IsOK()) {
+        THROW_ERROR_EXCEPTION(
+            NQueryTrackerClient::EErrorCode::QueryNotFound,
+            "Query %v is not found neither in active nor in finished query tables",
+            queryId)
+            << error;
+    }
+    bool isActive = asyncActiveRecord.IsSet() && asyncActiveRecord.Get().IsOK();
+    bool isFinished = asyncFinishedRecord.IsSet() && asyncFinishedRecord.Get().IsOK();
+    YT_VERIFY(isActive || isFinished);
+    if (isActive && isFinished) {
+        const auto& Logger = logger;
+        YT_LOG_ALERT(
+            "Query is found in both active and finished query tables "
+            "(QueryId: %v, Timestamp: %v)",
+            queryId,
+            timestamp);
+    }
+    TQuery result;
+    if (isActive) {
+        result = PartialRecordToQuery(asyncActiveRecord.Get().Value());
+    } else {
+        result = PartialRecordToQuery(asyncFinishedRecord.Get().Value());
+    }
+    return result;
+}
+
+void CheckAccessControlByQueryId(
+    TQueryId queryId,
+    const TString& root,
+    NTransactionClient::TTimestamp timestamp,
+    const TString& user,
+    IClientPtr client,
+    EPermission permission,
+    const NLogging::TLogger& logger)
+{
+    std::vector<TString> lookupKeys = {"user", "access_control_object"};
+    auto query = LookupQuery(queryId, client, root, lookupKeys, timestamp, logger);
+    if (CheckAccessControl(user, query.AccessControlObject, *query.User, client, permission) == NSecurityClient::ESecurityAction::Deny) {
+        ThrowAccessDeniedException(queryId, permission, user, query.AccessControlObject, *query.User);
+    }
+}
+
+std::set<TString> GetUserSubjects(TString user, IClientPtr client) {
+    // Get all subjects for the user.
+    TGetNodeOptions getFromCacheOptions;
+    getFromCacheOptions.ReadFrom = EMasterChannelKind::Cache;
+    auto userSubjectsOrError = WaitFor(client->GetNode("//sys/users/" + user + "/@member_of_closure", getFromCacheOptions));
+    if (!userSubjectsOrError.IsOK()) {
+        THROW_ERROR_EXCEPTION("Error while fetching @member_of_closure for the user @Qv", user)
+            << userSubjectsOrError;
+    }
+    auto userSubjects = ConvertToNode(userSubjectsOrError.Value())->AsList()->GetChildren();
+    std::set<TString> userSubjectsSet;
+    for (const auto& subject : userSubjects) {
+        userSubjectsSet.insert(subject->GetValue<TString>());
+    }
+    return userSubjectsSet;
+}
+
+std::vector<TString> GetACOsForSubjects(const std::set<TString>& subjects, IClientPtr client) {
+    // Get all access control objects.
+    TGetNodeOptions getPrincipalACLsFromCacheOptions;
+    getPrincipalACLsFromCacheOptions.Attributes = {
+        "principal_acl",
+    };
+    getPrincipalACLsFromCacheOptions.ReadFrom = EMasterChannelKind::Cache;
+
+    auto allACOsOrError = WaitFor(client->GetNode("//sys/access_control_object_namespaces/queries", getPrincipalACLsFromCacheOptions));
```

Comment:
Are you sure that fetching all ACOs from master is cheaper than doing batch "check_permissions" query?

Reply:
Discussed with Maxim: these ACLs can be cached on master caches and check_permissions cannot be cached.